ulog_cpp/messages.hpp

```diff
@@ -46,44 +56,392 @@ class FileHeader {
+   * In case the underlying type is a char array, it will be extracted as a std::string.
+   * In case the underlying type is an array, but we access a specific element, it will be
+   * extracted as the basic type T.
+   * @return std::variant of the value, in type defined in the Message Format
+   */
+  NativeTypeVariant asNativeTypeVariant() const;
+
+  /**
+   * @brief Cast the value to a desired, compile time known type. This will perform a static_cast
+   * to the desired type, if the type conversion is allowed by static_cast.
+   * For example one can extract a value defined as a uint8_t as an int.
+   *
+   * There are the following rules:
+   * - If requested type is vector<T> and native type is also vector<T>, vector<T> is returned
+   * - If requested type is vector<T>, but native type is vector<K>, each element gets casted from K
+   * to T
+   * - If requested type is not a vector, but native type is a vector, the first element is returned
+   * - If requested type is a vector, but native type is not a vector, a vector with 1 element is
+   * returned
+   * - If requested type is T and native type is K, a static_cast is performed on the element
+   * - If requested type is T and native type is T as well, the element is returned
+   *
+   * Specifically, when casting (signed) char, char is first casted to unsigned char, as this
+   * is generally the more expected behaviour (e.g. when casting a char to an int, one would
+   * expect the ASCII value of the char, not the signed value).
+   * @tparam T
+   * @return
+   */
+  template <typename T>
+  T as() const
+  {
+    T res;
+    std::visit(
+        [&res](auto&& arg) {
+          using NativeType = std::decay_t<decltype(arg)>;
+          using ReturnType = T;
+          if constexpr (is_string<NativeType>::value || is_string<ReturnType>::value) {
+            // there is a string involved. Only allowed when both are string
+            if constexpr (is_string<NativeType>::value && is_string<ReturnType>::value) {
+              // both are string
+              res = arg;
+            } else {
+              // one is string, the other is not
+              throw ParsingException("Assign strings and non-string types");
+            }
+          } else if constexpr (is_vector<NativeType>::value) {
+            // this is natively a vector
+            if constexpr (is_vector<ReturnType>::value) {
+              // return type is also vector
+              if constexpr (std::is_same<typename NativeType::value_type,
+                                         typename ReturnType::value_type>::value) {
+                // return type is same as native type
+                res = arg;
+              } else {
+                // return type is different from native type, but a vector
+                res.resize(arg.size());
+                for (std::size_t i = 0; i < arg.size(); i++) {
+                  res[i] = static_cast<typename ReturnType::value_type>(arg[i]);
```

Comment:
We can never hit this case, as here we'd have a char array (native) and another numeric array as a return type. This would error out before, as char arrays are handled as strings. 

This is also why clang-tidy never complained here, as this code path can not be hit with a signed char as the argument.

Reply:
Changed it anyway for readability